blog/_posts/2021-08-20-more-prefab-tools.md

```diff
@@ -0,0 +1,177 @@
+So replacing the stream operator for `Card` had the effect that when the constructor tried to add the body div, it tried to put the body div inside _itself_.
+
+![A gif of zooming in on cup with a small table next to it with a cup with a small table next to it with...]({{ site.baseurl }}/assets/lemniscate8/more-prefab-tools-droste.gif){:style="width: 100%"}
+*TOO MUCH RECURSION!!!*
+
+Using `static_cast<Div>(*this)` prevents infinite recursion by using `Div`'s default method of streaming components.
+
+#### The Config Panel Conundrum
+The next issue we tackled was a strange bug with the `ConfigPanel`.
+
+:bug: __Bug__: A `ConfigPanel` only works if it is a global object.
+
+This indicated that something was going wrong with our memory management.
+The problem was a rather subtle one, while the `ConfigPanel` _was_ a div, it also _had_ a div.
+Like Dr. Jekyll and Mr. Hyde caught mid transformation the `ConfigPanel` was caught trying to be two things at once, a class to setup a div :hammer_and_wrench: and a div with some extra frills :ribbon:.
+The result was that when time came to add the component to the document object model (DOM), the inner shell member div was properly placed and the outer shell destructed, but the inner shell kept references to the outer.
+
+:exclamation: And since the outer shell didn't exist at that point things were seriously borked. :exclamation:
+
+Keeping the component global kept the outer shell in existence past the end of `main` which is why the component could function.
+Once the problem was identified it wasn't to hard to fix its identity crisis.
+To align its behavior with other components, I eliminated the internal div and made sure users have to stream the entire component into the document.
+I also made sure that lambda expression capture by copy any components they need to keep around.
+Since Empirical web components are really just interfaces around a protected shared pointer this keeps access to the shared pointer with little overhead and prevents dangling references inside the lambda.
+In addition, the alterations provided the opportunity to break the component into some smaller reusable pieces.
+
+![A "refactor all the things" meme]({{ site.baseurl }}/assets/lemniscate8/more-prefab-tools-refactor.png){:style="width: 100%"}
+*I had to be a little careful not get too carried away*
+
+#### ReadoutPanel: Live Variables in the Limelight
+As a counterpart to the configuration panel, Matthew suggested making a component that could display values live from a simulation.
+While Empirical has a visualization component using [D3](https://d3js.org/), this component would serve the need for a lightweight display for simulation statistics.
+
+![Gollum eating a fish]({{ site.baseurl }}/assets/lemniscate8/more-prefab-tools-raw.png){:style="width: 50%"}
+*If you like your numbers like Gollum likes his fish, the `ReadoutPanel` might be for you*
```

Comment:
I like all the gifs and fun things you have in your post, I'd like to add some to ours!